core/src/main/java/org/apache/druid/data/input/impl/TimedShutoffInputSourceReader.java

```diff
@@ -68,8 +68,10 @@ public TimedShutoffInputSourceReader(InputSourceReader delegate, DateTime shutof
   )
   {
     final Closer closer = Closer.create();
-    closer.register(delegateIterator);
+    // We must register the shutdownNow method of the shutdown executor first. This will cause the shutdown executor
+    // to be closed last and ensure other resources which depend on it can close successfully.
     closer.register(exec::shutdownNow);
```

Comment:
Hmm, a few comments about this:

1. It's kind of poor form to close an executor from within a thread running in that same executor, which is what's happening here. Switching the order is a bit better, but it's still sort of like the picture below.
2. The record supplier's close function doesn't protect against multiple closes, and neither do Closers, so the record supplier might be closed multiple times or even concurrently. We should include something to ensure it will be OK if that happens.
3. Any ideas about how to test this? We should have testing for this sort of bug.

---

A photo of an executor being closed from a thread inside that same executor:

![image](https://user-images.githubusercontent.com/1214075/94980018-203ab780-04db-11eb-857b-16a72fb500cf.png)

Reply:
1. Removed exec::shutdownNow from the closer in the TimedShutoffInputSourceReader#decorateShutdownTimeout. Added a shutdown on the exec instead to make sure that after the timeout period, the exec will automatically be shutdown. This remove the dependency of shutting down the exec from the CloseableIterator#close()
2. Made the recordSuppliers close function thread safe by adding synchronize block. Left the Closers the same as it is also a documented feature/bug on Guava's Closer it is not thread safe (https://github.com/google/guava/issues/3900). We can keep this the same to avoid confusion and possible update our Closer copy when the above issue is done.
The CloseableIterator returned by TimedShutoffInputSourceReader#decorateShutdownTimeout also protect against concurrent close (scheduled exec vs normal closing) by using AtomicBoolean.

Reply:
3. Added tests for the following cases:
- Main thread close the iterator using the close() method (iterator successfully closed before timeout)
- Timeout close the iterator (iterator successfully closed by timeout before the close() of the iterator is called in the main thread)
- Timeout close the iterator and then while it is still closing, main thread close the iterator using the close() method
- Main thread close the iterator using the close() method and then while it is still closing, timeout close the iterator.